[RLlib] Learner API: Fix and unify grad-clipping configs and behaviors.

sven1977 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

Learner API: Fix and unify grad-clipping configs and behaviors.
* Introduce 3 new AlgorithmConfig settings: `grad_clip_by_value`, `grad_clip_by_norm` and `grad_clip_by_global_norm`.
* Implement proper clipping behaviors based on these 3 new settings in Tf- and TorchLearner (postprocess_gradients method).
* Add test cases for these new behaviors.

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

avnishn commented on rllib/core/learner/tf/tf_learner.py:
```diff
@@ -96,12 +97,49 @@ def compute_gradients(
+            }
+
+        # Clip by L2-norm (per gradient tensor).
+        grad_clip_by_norm = self._optimizer_config.get("grad_clip_by_norm", None)
```
are we going to allow users to clip gradients by value and then by norm? that doesn't sound right to me, but that is the behavior that has been enabled here.

avnishn commented on rllib/algorithms/qmix/qmix.py:
```diff
@@ -77,7 +77,10 @@ def __init__(self):
+
+        # TODO (sven): Deprecate grad_clip setting once all-in on new Learner API.
+        self.grad_clip = 10.0
+        self.grad_clip_by_global_norm = 10.0
```
how is this going to be used downstream by the qmix policy / multi_gpu_train_one_batch.

sven1977 replied:
fixed by replacing the new settings with backward-compatible ones

avnishn reviewed (commented):
Thanks for sharing this. However, I am overall confused by this pr. Is it incomplete?

sven1977 commented on rllib/core/learner/tf/tf_learner.py:
```diff
@@ -96,12 +97,49 @@ def compute_gradients(
+                ) for k, v in gradients_dict.items()
+            }
+
+        # Clip by L2-norm (per gradient tensor).
```
Great point. Was thinking about this myself. Maybe we should just do: `grad_clip: [some value]` and then `grad_clip_by: [value|norm|global_norm]`.

sven1977 commented:
Hey @avnishn , sorry about the confusion and thanks for taking a look!

I went through the PR once more and your comments and addressed all of them.
I changed the configs to the backward-compatible pair of `grad_clip` and (new) `grad_clip_by`, which specifies the mode of clipping. This way, it's a) backward compatible and b) exclusive (only one way of clipping at the same time allowed).

sven1977 commented on rllib/algorithms/impala/impala.py:
```diff
@@ -422,16 +425,6 @@ def validate(self) -> None:
             self.vtrace_clip_pg_rho_threshold
         )
 
-    @override(AlgorithmConfig)
```
Not needed anymore with this PR. grad clipping has been universally moved into `Learner.postprocess_gradients()`.

sven1977 commented on rllib/core/learner/torch/torch_learner.py:
```diff
@@ -79,6 +79,22 @@ def compute_gradients(
 
         return grads
 
+    @override(Learner)
```
This was missing in torch thus far.

sven1977 commented on rllib/utils/tf_utils.py:
```diff
@@ -27,6 +27,48 @@
 tf1, tf, tfv = try_import_tf()
 
 
+@PublicAPI
```
New grad clip utilities, replacing the old, messy ones (some of the old ones do clipping by value, some by norm (not global norm), some have additional optimizer update logic in them, etc..
